### PR TITLE
Fix ptype2 inheritance

### PR DESCRIPTION
**Pull request by lionel-**
Branched from #942.
Part of #710.

`vec_ptype2()` now determines methods from the first component of the class vector. This disables inheritance, which methods no longer need to guard from.

### CODE REVIEW

**Comment by DavisVaughan** [comment]
Just a comment as I'm reading through. Most R level `vec_ptype2()` methods are now only ever used for shaped input. So we will need to implement that at the C level too so we can eventually make the R level methods "dummy" methods

File: R/type-list-of.R [hunk 1]
```diff
@@ -185,11 +185,7 @@ as.character.vctrs_list_of <- function(x, ...) {
 #' @method vec_ptype2 vctrs_list_of
 #' @export
 vec_ptype2.vctrs_list_of <- function(x, y, ..., x_arg = "x", y_arg = "y") {
-  if (inherits_only(x, c("vctrs_list_of", "vctrs_vctr"))) {
-    UseMethod("vec_ptype2.vctrs_list_of", y)
-  } else {
-    vec_default_ptype2(x, y, x_arg = x_arg, y_arg = y_arg)
-  }
+  UseMethod("vec_ptype2.vctrs_list_of", y)
```
Comment (DavisVaughan) [inline R/type-list-of.R]: I guess you don't need `y` here now right?

Comment (DavisVaughan) [inline R/type-list-of.R]: I guess you don't need `y` here now right?

File: R/type2.R [hunk 2]
```diff
@@ -23,6 +23,27 @@
 #' return the same value as `vec_ptype2.y.x()`; this is not enforced
 #' for reasons of efficiency, but should be tested.
 #'
+#' `vec_ptype2()` are not inherited, classes must explicitly implement
+#' the methods. There are two reasons for this:
```
Comment (DavisVaughan) [inline R/type2.R]:
```suggestion
#' `vec_ptype2()` methods are not inherited, classes must explicitly implement
#' them. There are two reasons for this:
```

File: tests/testthat/test-type2.R [hunk 3]
```diff
@@ -158,11 +158,11 @@ test_that("Subclasses of data.frame dispatch to `vec_ptype2()` methods", {
   expect_identical(vec_ptype2(mtcars, quux), "dispatched!")
 })
 
-test_that("Subclasses of `tbl_df` have `tbl_df` common type (#481)", {
+test_that("Subclasses of `tbl_df` do not have `tbl_df` common type (#481)", {
```
Comment (DavisVaughan) [inline tests/testthat/test-type2.R]: Maybe update or remove the issue number so we aren't confused if we look back on this?

```suggestion
test_that("Subclasses of `tbl_df` do not have `tbl_df` common type (#710)", {
```

File: R/type-list-of.R (hunk 1 above)
Comment (lionel-) [reply R/type-list-of.R]: Right

**Comment by lionel-** [comment]
> Most R level vec_ptype2() methods are now only ever used for shaped input. So we will need to implement that at the C level too so we can eventually make the R level methods "dummy" methods

That'd be great. I was wondering if we could include shape-matching as part of our default handling, even for foreign S3 classes.

File: R/type-list-of.R (hunk 1 above)
Comment (lionel-) [inline R/type-list-of.R]:
```suggestion
  UseMethod("vec_ptype2.vctrs_list_of")
```

File: NEWS.md [hunk 4]
```diff
@@ -1,6 +1,12 @@
 
 # vctrs (development version)
 
+* Double-dispatch methods for `vec_ptype2()` are no longer inherited
```
Comment (hadley) [inline NEWS.md]: I think this needs an example just to make this concrete

File: R/type-bare.R [hunk 5]
```diff
@@ -7,77 +7,49 @@
 #' @method vec_ptype2 logical
 #' @export
 vec_ptype2.logical <- function(x, y, ..., x_arg = "", y_arg = "") {
-  if (is.object(x)) {
-    vec_default_ptype2(x, y, x_arg = x_arg, y_arg = y_arg)
-  } else {
-    UseMethod("vec_ptype2.logical")
-  }
+  UseMethod("vec_ptype2.logical")
```
Comment (hadley) [inline R/type-bare.R]: Can you add a note in the vignette saying that these functions are never actually called?

File: src/ptype2-dispatch.c [hunk 6]
```diff
@@ -116,26 +116,17 @@ SEXP vec_ptype2_dispatch_s3(SEXP x,
...
-                                    x_table,
-                                    &y_method_sym);
-
-  // FIXME: The `AsIs` class relies on a default ptype2 method. Should
-  // be changed with a general mechanism for wrapper types.
-  if (y_method == R_NilValue && Rf_inherits(x, "AsIs")) {
-    y_method_sym = s3_paste_method_sym(x_method_str, "default");
-    y_method = s3_sym_get_method(y_method_sym, x_table);
-  }
+  SEXP y_method = PROTECT(get_ptype2_method(y,
```
Comment (DavisVaughan) [inline src/ptype2-dispatch.c]: Nice!

**Pull request merged**